lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,383 @@
+  analysisUsage.setPreservesCFG();
+}
+
+// =====================================================================================================================
+// Lift readfirstlanes in relevant basic blocks to transform VGPR instructions into SGPR instructions as many as
+// possible.
+//
+// @param [in,out] function : LLVM function to be run for readfirstlane optimization.
+bool PatchReadFirstLane::liftReadFirstLane(Function &function) {
+  // Collect the basic blocks with amdgcn_readfirstlane
+  // Build the map between initial readfirstlanes and their corrensponding blocks
+  DenseMap<BasicBlock *, DenseSet<Instruction *>> blockInitialReadFirstLanesMap;
+  Module *module = function.getParent();
+  for (auto &func : *module) {
+    if (func.getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      for (User *user : func.users()) {
+        Instruction *inst = cast<Instruction>(user);
+        if (inst->getFunction() != &function)
+          continue;
+        blockInitialReadFirstLanesMap[inst->getParent()].insert(inst);
+      }
+      break;
+    }
+  }
+  bool changed = false;
+
+  // Lift readfirstlanes in each relevant basic block
+  for (auto blockInitialReadFirstLanes : blockInitialReadFirstLanesMap) {
+    BasicBlock *curBb = blockInitialReadFirstLanes.first;
+
+    // Step 1: Collect all instructions that "can be assumed uniform" with its divergent uses in a map
+    // (m_canAssumeUniformDivergentUsesMap)
+    collectAssumeUniforms(curBb, blockInitialReadFirstLanes.second);
+
+    // Step 2: Determine the best places to insert readfirstlane according to a heuristic
+    findInsertLocations(curBb);
+
+    // Step 3: Apply readFirstLane on all determined locations
+    if (!m_insertLocations.empty()) {
+      BuilderBase builder(curBb->getContext());
+      for (auto inst : m_insertLocations) {
+        // Avoid to insert reduncant readfirstlane
+        if (areAllUserReadFirstLane(inst))
+          continue;
+        applyReadFirstLane(inst, builder);
+        changed = true;
+      }
+    }
+  }
+  return changed;
+}
+
+// =====================================================================================================================
+// Collect the instructions that can be "assumed uniform" if the value itself isn't known to be uniform (by
+// DivergenceAnalysis), but we can replace all of its uses by readfirstlane(V) and still be correct
+//
+// @param block : The processing basic block
+// @param initialReadFirstLanes : The initial amdgcn_readfirstlane set
+void PatchReadFirstLane::collectAssumeUniforms(BasicBlock *block, DenseSet<Instruction *> &initialReadFirstLanes) {
+  DenseSet<Instruction *> canAssumeUniforms = std::move(initialReadFirstLanes);
+
+  for (BasicBlock::reverse_iterator instIt = block->rbegin(), endIt = block->rend(); instIt != endIt; ++instIt) {
+    Instruction *inst = &*instIt;
+    // Skip readfirstlane-irrelevant instructions
+    if (canAssumeUniforms.count(inst) == 0)
+      continue;
```

Comment:
Should the isSourceOfDivergence check go here instead of in `isBestInsertLocation`? The argument would be: If instruction I can be assumed uniform, but it is a source of divergence, then it's not clear that its operands can also automatically be assumed to be uniform.